compiler/noirc_evaluator/src/ssa/opt/remove_unreachable.rs

```diff
@@ -0,0 +1,75 @@
+use std::collections::BTreeSet;
+
+use fxhash::FxHashSet as HashSet;
+
+use crate::ssa::{
+    ir::{
+        function::{Function, FunctionId},
+        instruction::Instruction,
+        value::Value,
+    },
+    ssa_gen::Ssa,
+};
+
+impl Ssa {
+    /// Removes any unreachable functions from the code. These can result from
+    /// optimizations making existing functions unreachable, e.g. `if false { foo() }`,
+    /// or even from monomorphizing an unconstrained version of a constrained function
+    /// where the original constrained version ends up never being used.
+    pub(crate) fn remove_unreachable_functions(mut self) -> Self {
+        let mut used_functions = HashSet::default();
+
+        for function_id in self.functions.keys() {
+            if self.is_entry_point(*function_id) {
+                collect_reachable_functions(&self, *function_id, &mut used_functions);
+            }
+        }
+
+        self.functions.retain(|id, _| used_functions.contains(id));
+        self
+    }
+}
+
+fn collect_reachable_functions(
+    ssa: &Ssa,
+    current_func_id: FunctionId,
+    reachable_functions: &mut HashSet<FunctionId>,
+) {
+    if reachable_functions.contains(&current_func_id) {
+        return;
+    }
+    reachable_functions.insert(current_func_id);
+
+    let func = &ssa.functions[&current_func_id];
+    let used_functions = used_functions(func);
+
+    for called_func_id in used_functions.iter() {
+        collect_reachable_functions(ssa, *called_func_id, reachable_functions);
+    }
+}
+
+fn used_functions(func: &Function) -> BTreeSet<FunctionId> {
+    let mut used_function_ids = BTreeSet::default();
+
+    let mut find_functions = |value| {
+        if let Value::Function(function) = func.dfg[value] {
+            used_function_ids.insert(function);
+        }
+    };
+
+    for block_id in func.reachable_blocks() {
+        let block = &func.dfg[block_id];
+
+        for instruction_id in block.instructions() {
+            let instruction = &func.dfg[*instruction_id];
+
+            if matches!(instruction, Instruction::Store { .. } | Instruction::Call { .. }) {
```

Comment:
Just for my understanding: what exactly does it mean to `Store` but never `Call` a function?

Reply:
If we store a function in another value, we may eventually load the function from that value and call it.

Reply:
That makes sense, I was just wondering if it's considered reachable even if it's never called, ie. if we only ever see a `Store` but never a `Call` for it.

Reply:
That could possibly occur and yes technically that function be an unreachable. But we want to leave any `Store`/`Load` resolution to mem2reg. If a `Store` remains we should assume that the value we are storing is potentially used elsewhere in a separate block or even passed as a reference to another function.

Reply:
Right, so removing a function that it referenced by something, even if not called, could be a problem. Thanks!